### ISSUE TITLE
[JENKINS-40966] Ability to leave Declarative node block w/ top-level agent for a single stage

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/pipeline-model-definition-plugin#928]
<p>I have a use case where I'd like to have a specific section of the pipeline limited to one concurrent step across all pipeline instances.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeHeader panelHeader" style="border-bottom-width: 1px;"><b>Jenkinsfile</b></div><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent {
    docker {
      label <span class="code-quote">'label'</span>
      image <span class="code-quote">'image'</span>
    }
  }

  stages { 
    stage(<span class="code-quote">'locked_staged'</span>) {
      steps {
<span class="code-comment">//<span class="code-keyword">do</span> some work
</span>lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
  <span class="code-comment">//<span class="code-keyword">do</span> locked work
</span>}
<span class="code-comment">//<span class="code-keyword">do</span> more work
</span>      }
    }
    
    stage(<span class="code-quote">'unlocked_stage'</span>) {
      steps {
<span class="code-comment">//work
</span>      }
    }
}
</pre>
</div></div>

<p>This example would consume a node for the entire pipeline. Any node hitting the lock stage would current be blocked waiting on the lock. I would like for the node to be released once the lock is hit and resume once available.</p>

<p>There is a partial workaround by specifically declaring agent at the appropriate stage.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeHeader panelHeader" style="border-bottom-width: 1px;"><b>Jenkinsfile with workaround</b></div><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent none

  stages { 
     stage(<span class="code-quote">'locked_staged'</span>) {
agent <span class="code-quote">'label'</span>
steps {
  lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
    <span class="code-comment">//<span class="code-keyword">do</span> some work
</span>  }
       }
    }
    
    stage(<span class="code-quote">'unlocked_stage'</span>) {
      steps {
<span class="code-comment">//work
</span>      }
    }
}
</pre>
</div></div>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpocase">rpocase</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40966" target="_blank">Ability to leave Declarative node block w/ top-level agent for a single stage</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: pipeline-model-definition-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 3
<li><b>watchers</b>: 7
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I have a use case where I'd like to have a specific section of the pipeline limited to one concurrent step across all pipeline instances.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeHeader panelHeader" style="border-bottom-width: 1px;"><b>Jenkinsfile</b></div><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent {
    docker {
      label <span class="code-quote">'label'</span>
      image <span class="code-quote">'image'</span>
    }
  }

  stages { 
    stage(<span class="code-quote">'locked_staged'</span>) {
      steps {
        <span class="code-comment">//<span class="code-keyword">do</span> some work
</span>        lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
          <span class="code-comment">//<span class="code-keyword">do</span> locked work
</span>        }
        <span class="code-comment">//<span class="code-keyword">do</span> more work
</span>      }
    }
    
    stage(<span class="code-quote">'unlocked_stage'</span>) {
      steps {
        <span class="code-comment">//work
</span>      }
    }
}
</pre>
</div></div>

<p>This example would consume a node for the entire pipeline. Any node hitting the lock stage would current be blocked waiting on the lock. I would like for the node to be released once the lock is hit and resume once available.</p>

<p>There is a partial workaround by specifically declaring agent at the appropriate stage.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeHeader panelHeader" style="border-bottom-width: 1px;"><b>Jenkinsfile with workaround</b></div><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent none

  stages { 
     stage(<span class="code-quote">'locked_staged'</span>) {
        agent <span class="code-quote">'label'</span>
        steps {
          lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
            <span class="code-comment">//<span class="code-keyword">do</span> some work
</span>          }
       }
    }
    
    stage(<span class="code-quote">'unlocked_stage'</span>) {
      steps {
        <span class="code-comment">//work
</span>      }
    }
}
</pre>
</div></div></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>pipeline-model-definition 0.8.1</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/hrmpw.png" title="hrmpw's avatar" /> hrmpw:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/40966?focusedId=282781&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282781">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40966?focusedId=282781&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282781">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>you can also use the <tt>node</tt> step directly in a <tt>stage</tt> to get an executor.</p>


<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent none
  stages { 
    stage(<span class="code-quote">'locked_staged'</span>) {
      steps {
        lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
           echo <span class="code-quote">"I'm locked"</span>
        }
        node(<span class="code-quote">'my-label'</span>) {
          echo <span class="code-quote">"I'm on a node and not locked"</span>
        }
      }
    }
  }
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>you can also use the <tt>node</tt> step directly in a <tt>stage</tt> to get an executor.</p>



<pre>
pipeline {
  agent none
  stages { 
    stage(<span class="code-quote">'locked_staged'</span>) {
      steps {
lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
   echo <span class="code-quote">"I'm locked"</span>
}
node(<span class="code-quote">'my-label'</span>) {
  echo <span class="code-quote">"I'm on a node and not locked"</span>
}
      }
    }
  }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/40966?focusedId=282785&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282785">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40966?focusedId=282785&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282785">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hmm - good point. With <tt>agent</tt> on the stage, the <tt>lock</tt> isn't actually happening until you've got the executor on the agent. So if you've only got one executor on that agent (or all the executors are filled up with pipelines, either locked or running), the build may end up waiting a while to even get the executor before locking, and you end up with the locked pipelines still eating an executor. Hrrrrm. We might want to add a section to <tt>stage</tt> like <tt>options</tt> at the top-level - you can run the entire build in things like <tt>timeout</tt> or <tt>timestamps</tt> there, before even the top-level <tt>agent</tt> is invoked. Might make sense to have the equivalent pre-<tt>agent</tt> behavior on <tt>stage</tt> too...</p>
</pre>
</details>

</ul>
</details>

<p>Hmm - good point. With <tt>agent</tt> on the stage, the <tt>lock</tt> isn't actually happening until you've got the executor on the agent. So if you've only got one executor on that agent (or all the executors are filled up with pipelines, either locked or running), the build may end up waiting a while to even get the executor before locking, and you end up with the locked pipelines still eating an executor. Hrrrrm. We might want to add a section to <tt>stage</tt> like <tt>options</tt> at the top-level - you can run the entire build in things like <tt>timeout</tt> or <tt>timestamps</tt> there, before even the top-level <tt>agent</tt> is invoked. Might make sense to have the equivalent pre-<tt>agent</tt> behavior on <tt>stage</tt> too...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rpocase:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/40966?focusedId=282929&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282929">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40966?focusedId=282929&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282929">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is it possible to lock before entering a node block? I would expect this to put the pipeline into flyweight while waiting on an executor and locked. Testing locally gave me a "No such DSL method 'lock' found among steps" - not sure if I'm missing an API extension plugin.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent none
  stages { 
    stage(<span class="code-quote">'locked_staged'</span>) {
      steps {
        node(<span class="code-quote">'my-label'</span>) {
          echo <span class="code-quote">"I'm on a node and not locked"</span>
        }
        lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
           echo <span class="code-quote">"I'm locked"</span>
           node(<span class="code-quote">'my-label'</span>) {
            echo <span class="code-quote">"I'm on a node and locked"</span>
          }
        }       
      }
    }
  }
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Is it possible to lock before entering a node block? I would expect this to put the pipeline into flyweight while waiting on an executor and locked. Testing locally gave me a "No such DSL method 'lock' found among steps" - not sure if I'm missing an API extension plugin.</p>


<pre>
pipeline {
  agent none
  stages { 
    stage(<span class="code-quote">'locked_staged'</span>) {
      steps {
node(<span class="code-quote">'my-label'</span>) {
  echo <span class="code-quote">"I'm on a node and not locked"</span>
}
lock(resource: <span class="code-quote">'resource'</span>, inversePrecedence:<span class="code-keyword">false</span>) {
   echo <span class="code-quote">"I'm locked"</span>
   node(<span class="code-quote">'my-label'</span>) {
    echo <span class="code-quote">"I'm on a node and locked"</span>
  }
}       
      }
    }
  }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/hrmpw.png" title="hrmpw's avatar" /> hrmpw:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/40966?focusedId=282931&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282931">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40966?focusedId=282931&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282931">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You need to install the <a href="https://wiki.jenkins-ci.org/display/JENKINS/Lockable+Resources+Plugin" class="external-link" target="_blank" rel="nofollow noopener">lockable-resources-plugin</a> to use <tt>lock</tt> in Pipeline.</p>

<p>In your example able the first node is not blocked but the second node would be locked. </p>
</pre>
</details>

</ul>
</details>

<p>You need to install the <a href="https://wiki.jenkins-ci.org/display/JENKINS/Lockable+Resources+Plugin" class="external-link" target="_blank" rel="nofollow noopener">lockable-resources-plugin</a> to use <tt>lock</tt> in Pipeline.</p>

<p>In your example able the first node is not blocked but the second node would be locked. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/40966?focusedId=291645&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291645">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40966?focusedId=291645&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291645">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>So I'm generalizing this ticket! There are multiple cases where we want to be able to "leave" the <tt>node</tt> block created at the top-level by use of anything but <tt>agent none</tt> but then "return" to it after a stage with a different <tt>agent</tt> (particularly <tt>agent none</tt>) has completed. There's <tt>lock</tt>, there's the <tt>throttle</tt> step I've been working on in <a href="https://issue-redirect.jenkins.io/issue/31801" title="Pipeline Support in Throttle Concurrent Builds" class="issue-link" data-issue-key="JENKINS-31801"><del>JENKINS-31801</del></a>, there's CloudBees' proprietary <tt>checkpoint</tt> feature, there's the <tt>input</tt> step, and there's probably more I'm not thinking of.</p>

<p>To do this, I think there are a few functional requirements:</p>
<ul>
	<li>Free up the original executor so that other builds can use it while we're not.</li>
	<li>Somehow "lock" the original workspace so that it doesn't get reused or deleted while we're not on that executor - i.e., we want to make sure another run of the same Pipeline doesn't end up reusing it and instead creates "somePipeline@2" etc.</li>
	<li>Ideally, keep the computer/node the original executor was on from being deleted due to <tt>RetentionStrategy</tt> until after we've come back to it and are truly done with it. This one is a very-nice-to-have rather than a hard requirement, though - you can work around this by using <tt>stash</tt> at the end of the <tt>stage</tt> before you leave the original executor and <tt>unstash</tt> at the beginning of the stage where you return to that agent/label/whatever. But <tt>stash</tt> is not very efficient, shall we say.</li>
	<li>In the same vein, make sure we know the specific node/computer the original executor was on so we can return there. However, this is only useful if we've managed to "lock" the workspace to prevent reuse/deletion <b>and</b> we've managed to guarantee the node/computer won't be automatically deleted via a <tt>RetentionStrategy</tt> before we come back to it. If we can't guarantee reusability of the exact workspace and continued existence of the exact node/computer, knowing exactly where we ran before isn't necessarily of any use.</li>
	<li>We need to decide whether we do this automatically whenever a per-stage <tt>agent</tt> is specified that does not include <tt>reuseNode true</tt> (an option on <tt>docker</tt> and <tt>dockerfile</tt> agent types) or if we require an additional configuration option (in the <tt>agent</tt> or on the <tt>stage</tt> itself) when using any agent but <tt>agent none</tt>. Usability is fun!</li>
</ul>


<p>So...yeah. I'm going to try to write up an RFC for this in the next few days, with the hope of getting that approved in the next week or two, followed by implementation in time for Declarative 1.2 (which is tentatively scheduled for late April).</p>
</pre>
</details>

</ul>
</details>

<p>So I'm generalizing this ticket! There are multiple cases where we want to be able to "leave" the <tt>node</tt> block created at the top-level by use of anything but <tt>agent none</tt> but then "return" to it after a stage with a different <tt>agent</tt> (particularly <tt>agent none</tt>) has completed. There's <tt>lock</tt>, there's the <tt>throttle</tt> step I've been working on in <a href="https://issue-redirect.jenkins.io/issue/31801" title="Pipeline Support in Throttle Concurrent Builds" class="issue-link" data-issue-key="JENKINS-31801"><del>JENKINS-31801</del></a>, there's CloudBees' proprietary <tt>checkpoint</tt> feature, there's the <tt>input</tt> step, and there's probably more I'm not thinking of.</p>

<p>To do this, I think there are a few functional requirements:</p>
<ul>
	<li>Free up the original executor so that other builds can use it while we're not.</li>
	<li>Somehow "lock" the original workspace so that it doesn't get reused or deleted while we're not on that executor - i.e., we want to make sure another run of the same Pipeline doesn't end up reusing it and instead creates "somePipeline@​2" etc.</li>
	<li>Ideally, keep the computer/node the original executor was on from being deleted due to <tt>RetentionStrategy</tt> until after we've come back to it and are truly done with it. This one is a very-nice-to-have rather than a hard requirement, though - you can work around this by using <tt>stash</tt> at the end of the <tt>stage</tt> before you leave the original executor and <tt>unstash</tt> at the beginning of the stage where you return to that agent/label/whatever. But <tt>stash</tt> is not very efficient, shall we say.</li>
	<li>In the same vein, make sure we know the specific node/computer the original executor was on so we can return there. However, this is only useful if we've managed to "lock" the workspace to prevent reuse/deletion <b>and</b> we've managed to guarantee the node/computer won't be automatically deleted via a <tt>RetentionStrategy</tt> before we come back to it. If we can't guarantee reusability of the exact workspace and continued existence of the exact node/computer, knowing exactly where we ran before isn't necessarily of any use.</li>
	<li>We need to decide whether we do this automatically whenever a per-stage <tt>agent</tt> is specified that does not include <tt>reuseNode true</tt> (an option on <tt>docker</tt> and <tt>dockerfile</tt> agent types) or if we require an additional configuration option (in the <tt>agent</tt> or on the <tt>stage</tt> itself) when using any agent but <tt>agent none</tt>. Usability is fun!</li>
</ul>


<p>So...yeah. I'm going to try to write up an RFC for this in the next few days, with the hope of getting that approved in the next week or two, followed by implementation in time for Declarative 1.2 (which is tentatively scheduled for late April).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>philip:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/40966?focusedId=301545&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-301545">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40966?focusedId=301545&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-301545">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> Any news on this one? Especially when using only one input step in a declarative pipeline if feels really cumbersome to introduce a node or agent for each stage to workaround this issue. The issue feels more like a bug or flaw in declarative than like a new feature request.</p>

<p>IMHO it's important that the fix provides not only the ability to leave the top-level node/agent for a single stage but for a single step within a stage. Otherwise you have to introduce an additional stage only to provide an input within declarative.</p>

<p>Example:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
    agent none
    stages {
        stage (<span class="code-quote">"test stage"</span>) {
            steps {
                input <span class="code-quote">"Continue?"</span>
                node(<span class="code-quote">"master"</span>) {
                    sh <span class="code-quote">'pwd'</span>
                }    
            }    
        }  
        stage (<span class="code-quote">"another stage"</span>) {
            steps {
                node(<span class="code-quote">"master"</span>) {
                    sh <span class="code-quote">'pwd'</span>
                    echo <span class="code-quote">'all other stages need node step :('</span>
                }    
            }    
        }
    }
}   
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> Any news on this one? Especially when using only one input step in a declarative pipeline if feels really cumbersome to introduce a node or agent for each stage to workaround this issue. The issue feels more like a bug or flaw in declarative than like a new feature request.</p>

<p>IMHO it's important that the fix provides not only the ability to leave the top-level node/agent for a single stage but for a single step within a stage. Otherwise you have to introduce an additional stage only to provide an input within declarative.</p>

<p>Example:</p>

<pre>
pipeline {
    agent none
    stages {
        stage (<span class="code-quote">"test stage"</span>) {
            steps {
                input <span class="code-quote">"Continue?"</span>
                node(<span class="code-quote">"master"</span>) {
                    sh <span class="code-quote">'pwd'</span>
                }    
            }    
        }  
        stage (<span class="code-quote">"another stage"</span>) {
            steps {
                node(<span class="code-quote">"master"</span>) {
                    sh <span class="code-quote">'pwd'</span>
                    echo <span class="code-quote">'all other stages need node step :('</span>
                }    
            }    
        }
    }
}   
</pre>
